_source/_posts/2022-08-19-build-crud-spring-and-vue.md

```diff
@@ -0,0 +1,1057 @@
+        this.$log.debug("Item removed:", todoToRemove);
+        this.todos.splice(this.todos.indexOf(todoToRemove), 1)
+      }).catch((error) => {
+        this.$log.debug(error);
+        this.error = "Failed to remove todo"
+      });
+    },
+
+    handleDeleteCompleted() {
+      const completed = this.todos.filter(todo => todo.completed)
+      Promise.all(completed.map( todoToRemove => {
+        return this.$api.removeForId(todoToRemove.id).then(() => {
+          this.$log.debug("Item removed:", todoToRemove);
+          this.todos.splice(this.todos.indexOf(todoToRemove), 1)
+        }).catch((error) => {
+          this.$log.debug(error);
+          this.error = "Failed to remove todo"
+          return error
+        })
+      }))
+    },
+
+    handleDoneEditingNewTodo() {
+      const value = this.newTodoTitle && this.newTodoTitle.trim()
+      if (!value) {
+        return
+      }
+      this.$api.createNew(value, false).then( (response) => {
+        this.$log.debug("New item created:", response)
+        this.newTodoTitle = ""
+        this.todos.push({
+          id: response.data.id,
+          title: value,
+          completed: false
+        })
+        this.$refs.newTodoInput.blur()
+      }).catch((error) => {
+        this.$log.debug(error);
+        this.error = "Failed to add todo"
+      });
+    },
+    handleCancelEditingNewTodo() {
+      this.newTodoTitle = ""
+    },
+
+    handleSetCompleted(id, value) {
+      let todo = this.todos.find(todo => id === todo.id)
+      todo.completed = value
+    },
+
+    handleSetTitle(id, value) {
+      let todo = this.todos.find(todo => id === todo.id)
+      todo.title = value
+    },
+
+    handleShowError(message) {
+      this.error = message
+    },
+
+    handleErrorClick () {
+      this.error = null;
+    },
+
+  },
+
+}
+</script>
+
+<style>
+#row-container {
+  margin-top: 100px;
+}
+.my-card {
+  min-width: 600px;
+}
+.error {
+  color: red;
+  text-align: center;
+  min-width: 600px;
+  margin-top: 10px;
+}
+</style>
+```
+{% endraw %}
+
+This component encapsulates the card that holds all of the todos, as well as the todo-associated interface elements. It also handles the rest of the functions related to updating todos on the server as well as in the local cache. 
+
+You're welcome to delete the `HelloWorld.vue` component, if you want. Or you can leave it. It's not needed.
+
+## Test the todo app
+
+Make sure the Spring Boot API is still running. In a separate Bash shell, from the resource server directory, run the following command (if it is not already still running)
+
+```bash
+./gradlew bootRun
+```
+
+Start the Vue app using the embedded development server. From the client directory:
+
+```bash
+npm run serve
+```
+
+Open a browser and navigate to `http://localhost:8080`. You'll see the "please log in" page.
+
+
+{% img blog/spring-boot-crud-update/please-log-in.png alt:"Please log in" width:"1000" %}{: .center-image }
```

Comment:
Mine looks slightly different on Chrome (104) and macOS Monterey (12.5.1).

<img width="657" alt="Screen Shot 2022-08-18 at 15 55 33" src="https://user-images.githubusercontent.com/17892/185502226-f67b9597-2c77-41cf-9d98-4d882adcaee6.png">